### PR TITLE
Implement GetStreamTimes API

### PR DESCRIPTION
**Pull request by phunkyfish**
4.5.0
- Implement GetStreamTimes API for timeshifting and recordings support
- Mac build script
- Fix compiler warnings

On recordings the Ui will correctly show the in progress recordings (red dot in Recordings view) and when viewing them the duration will increase as you watch.

### CODE REVIEW

**Comment by phunkyfish** [comment]
Sure, can drop reformatting.

It was broken for Mac before kodi 18 release. Fixed in 19 I think but still a useful script if you jump between addons like I do.

**Comment by phunkyfish** [comment]
Ok, should be ready for review now.

File: src/client.cpp [hunk 1]
```diff
@@ -518,10 +537,55 @@ extern "C" {
...
-    strncpy(properties[0].strValue, (*recIt)->m_url.c_str(), sizeof(properties[0].strValue) - 1);
-    *iPropertiesCount = 1;
-    return PVR_ERROR_NO_ERROR;
+    std::time_t now = std::time(nullptr), start = 0, end = 0;
+    std::string channelName = recording.strChannelName;
+    time_t recordingTime = recording.recordingTime;
+    auto timerIt = std::find_if(recordings.begin(), recordings.end(),
+      [now, channelName, recordingTime](const RecordingPtr &item)
+    {
+      return item->IsTimer() && item->IsRunning(now, channelName, recordingTime);
```
Comment (Jalle19) [inline src/client.cpp]: `IsRunning` already return false if it's not a timer, so technically the `IsTimer` check isn't needed here

**Comment by Jalle19** [comment]
Looks good in general, I take it the `RecordingReader` class is copied from some other addon?

File: src/client.cpp (hunk 1 above)
Comment (phunkyfish) [reply src/client.cpp]: Or remove the timer check from IsRunning as I guess it applies in either case.

**Comment by phunkyfish** [comment]
Ya, comes from vuplus, prior to that dvbviewer .

File: src/client.cpp (hunk 1 above)
Comment (Jalle19) [reply src/client.cpp]: Sure, that works too

**Comment by Jalle19** [comment]
> Ya, comes from vuplus, prior to that dvbviewer wink

I suspect there's a lot of code that could be factored out to a "libkodipvr" library. Local timeshift buffering, XMLTV parser, this recording reader (which is very similar in concept to the timeshift buffer used in this addon) etc.

**Comment by ksooo** [comment]
That would be cool, yes.

File: src/client.cpp (hunk 1 above)
Comment (Jalle19) [reply src/client.cpp]: Change this if you want to, I'll approve the PR nevertheless.

**Comment by Jalle19** [comment]
Noticed that AppVeyor integration is missing for this addon.

**Comment by phunkyfish** [comment]
Ya, I’ve been thinking the same. I’ve always wanted to build a proper circular buffer (memory/disk) for time shifting. One day.

**Comment by Jalle19** [comment]
> Ya, I’ve been thinking the same. I’ve always wanted to build a proper circular buffer (memory/disk) for time shifting. One day.

Yeah, that's one of the major limitations of this addon's current implementation (no circularity)

**Comment by phunkyfish** [comment]
I can add appveyor too.

**Pull request merged**